### DISCUSSION TITLE
v.2.2.000

### DISCUSSION THREAD

**Discussion by Medard22** [Medard22/Dolphin-MMJR2-VBI#88] [General]
I think it's finally time to stop tinkering and finally release a new version. So here it goes 😄 

**Changelog from v2.1.010 to v2.2.000**

- updated NDK to 25.1.8937393
- updated Android libraries
- Java 11 bytecode is now enabled across project
- code cleanup and refactoring
- minor fixes to JitArm64
- unitialized XFB color changed from pink/magenta/fuchsia to black, making rendering artifacts with "Store XFB copies to Texture only" less noticeable
- exposed "Store XFB Copies to Texture only" speedhack in Quick settings menu
- exposed "Show FPS" in the Quick settings menu
- fixed misaligned OSD messages
- hopefully fixed a throttle bug that made game run at super fast speed after loading state
- removed Autoupdater from the UI, it wasn´t working anyway
- added an option to force linear texture filtering and exposed it in the Android settings
- added sharp bilinear postprocessing shader
- added integer scaling shader
- implemented PPC Data Cache (currently not exposed in Android UI - kills performance)
- implemented CPU Vertex Culling (Graphics Settings->Advanced) - auto-enabled in Metroid Prime and Twilight Princess
- Skylanders Portal support is now in the core (but not exposed in Android due to missing Android UI)
- less confusing versioning
- other minor fixes and changes

**Known issues**
- When updating from older build, Dolphin may hang on building shader cache.
- If that happens delete shader cache folder from the user folder, disable "Compile Shaders Before Starting", run the game, exit emulator and then enable it again.

**Caution**

- **Save States from previous versions of MMJR2-VBI are NOT compatible! Make an in-game save, or you will lose progress!**
- This release will replace any MMJR2 build you might have installed! If the installation fails, uninstall your MMJR2 first. 
- Updating to this release will keep your original MMJR2 data intact. This release uses a separate /mmjr2-vbi folder in the root of your Internal storage.
- This release is incompatible with savestates from other forks (MMJR, MMJR2, Dolphin for Handheld) or Official Dolphin. 
- In-game saves are compatible. You can copy data to respective folders from your Dolphin user directory to the /mmjr2-vbi folder
- I am not an officially registered Android developer. Google Play Protect might get suspicious and give you a warning :)

**Additional information**

- For Daijishō users - this build uses **"106 - gc - Dolphin mmjr"** and **"334 - wii - Dolphin mmjr"** player settings
- All code ported over from Dolphin Official has links to original pull requests and respective authors. I deeply respect their work.
- This is a personal project. Additional updates and releases are possible but not guaranteed. If you enjoy it, great! Otherwise get Official Dolphin from PlayStore.

This release codebase is roughly equivalent to the Official Dolphin dev build 5.0-18439, Jan 30 2023. Don´t expect any massive performance gains, but there are some nice quality-of-life changes. 

Any Android developer willing to help me is welcome. Hell any developer willing to help is welcome 🙈  
The MMJR2 code and especially it´s Android part could use some love. Dolphin went through several massive code overhauls in 2023 and I am not sure how much they break MMJR2. I will try to keep up as long as my limited abilites let me. 

Feedback is welcome in the Discussions. You can also find me on Retroid Discord server. Happy gaming! ✌️ 

<hr /><em>This discussion was created from the release <a href='https://github.com/Medard22/Dolphin-MMJR2-VBI/releases/tag/'>v.2.2.000</a>.</em>

**Comment by ajay9634** [comment]
Hi

Can you fix IR drag function in mmjr 2 vbi?
There is no problem in official.

**Reply by ajay9634** [reply]
Also no problem in mmjr , only problem in mmjr 2.

IR Mode 'drag' function is working the same as 'follow'.

**Reply by Medard22** [reply]
Hi, sorry for the late reply. I was on a holiday over the Christmas. I can certainly look into it. I have never used this function, so could you please provide me with more information? What game are you playing that use this? What is the difference between the modes? I have to test it out myself.

**Reply by ajay9634** [reply]
Hi
I can explain through video.
In official - ir mode is working properly
But in mmjr 2 - ir mode is not working properly. Follow and drag function are working same in mmjr 2 . Which shouldn't do this.

This problem is related to the previous original fork of mmjr 2 but I couldn't contact the developers.

My device is poco M2 pro (snapdragon 720g, android 12.

Video link - (GitHub couldn't upload due to large size.)

https://mega.nz/folder/SzhCnZRK#MMasBjRrKLOcWgUXwQinBQ

**Reply by ajay9634** [reply]
Without ir fix, I can't play metroid prime 3 and other fps games

**Reply by Medard22** [reply]
Gotcha. It´s pretty much direct vs relative cursor emulation. I found the code responsible and I am looking into it. It seems like the IR pointer recenter is somehow always on. I will keep you posted. Now I need Metroid 3 to test it 🤔

**Reply by ajay9634** [reply]
The recenter option is off,
It's  like , codes are the same in drag mod and follow mod.

**Reply by Medard22** [reply]
Yeah the option is off, but somehow it behaves like it´s always on.

**Reply by Medard22** [reply]
Phew, finally found it. There was an old code, I guess from MMJR in the function that sets the IR mode. In MMJR it worked, but later in MMJR2 lifecycle, other changes was ported over. That section was however not updated and that´s why it was not working properly. I tested it on Metroid 3 and it looks fixed. Feel free to give it a spin. Here is a link to my Google Drive with an updated APK, based on top of the latest dev branch. AFAIK this update should not affect savestates from v.2.2.000 but make an in-game save just in case. 

[MMJR2-VBI-dev-IR-Hotfix.apk](https://drive.google.com/file/d/1XcOogFt6K9oBKeiaa5Ias7_Fen4ooQT6/view)

Thanks for letting me know!❤️
This could help a lot of people, so I will try to make a new updated release soon. Until then feel free to share that link.

**Comment by scgpgh** [comment]
I've been using this with my Retroid Pocket 2S and all I can say is great work!

**Comment by Horonelius** [comment]
This is fantastic, and the only thing keeping me from using it over regular Dolphin is that I can't figure out a way to designate stick sensitivity/range/multiplier. Twilight Princess runs great on my RP4 without any stutters, but the range between walk and run are practically nonexistant.

**Reply by Medard22** [reply]
There is no option in the GUI to control analog sensitivity, however there seems to be some builtin legacy option to tweak it via INI file. First off make a custom gamepad configuration for Twilight Princess. Then you can open that ini file and add Main Stick/Calibration = 133  That should supposedly make the analog stick less sensitive. I am not sure if this is the proper syntax though. This old system has been replaced around October 2022, try to google information between years 2019 and 2022. 

Source from here - https://forums.dolphin-emu.org/Thread-radius-settings-ignored-in-custom-profiles and here https://www.reddit.com/r/OdinHandheld/comments/ukqpmk/mmjr_joystick_sensitivity/

**Reply by Horonelius** [reply]
Thank's a million for guidance, my google-fu is not particularly strong. EDIT: THANK YOU SO MUCH!

**Reply by Medard22** [reply]
Test it out and let me know if it works 👍

**Reply by Horonelius** [reply]
Works like a charm! Thanks again!

**Reply by Medard22** [reply]
Great, good to know ^^b

**Reply by Horonelius** [reply]
The entire issue is that all documentation I was able to find still uses _Radius_ instead of _Calibration_, and then said to use  _Profile1_ instead instead of _PadProfile1_. I feel like this should all be written down somewhere.

**Reply by Medard22** [reply]
Can you send me your working ini? Maybe over Discord? I would like to have a look. As you said there isn't much info about this available. The input in mmjr2 will most likely remain untouched for the foreseeable future. I will add the this information to the release notes.
